system/lib/wasmfs/syscalls.cpp

```diff
@@ -535,4 +535,72 @@ __wasi_errno_t __wasi_fd_fdstat_get(__wasi_fd_t fd, __wasi_fdstat_t* stat) {
   }
   return __WASI_ERRNO_SUCCESS;
 }
+
+long __syscall_getdents64(long fd, long dirp, long count) {
+  auto openFile = wasmFS.getLockedFileTable()[fd];
+
+  if (!openFile) {
+    return -EBADF;
+  }
+  dirent* result = (dirent*)dirp;
+
+  // Check if the result buffer is too small.
+  if (count / sizeof(dirent) == 0) {
+    return -EINVAL;
+  }
+
+  auto file = openFile.locked().getFile();
+
+  auto directory = file->dynCast<Directory>();
+
+  if (!directory) {
+    return -ENOTDIR;
+  }
+
+  // Hold the locked directory to prevent the state from being changed during
+  // the operation.
+  auto lockedDir = directory->locked();
+
+  off_t bytesRead = 0;
+  // A directory's position corresponds to the index in its entries vector.
+  int index = openFile.locked().position();
+
+  // In the root directory, ".." refers to itself.
+  auto dotdot =
+    file == wasmFS.getRootDirectory() ? file : lockedDir.getParent();
+
+  // If the directory is unlinked then the parent pointer should be null.
+  if (!dotdot) {
+    return -ENOENT;
+  }
+
+  // There are always two hardcoded directories "." and ".."
+  std::vector<Directory::entry> entries = {{".", file}, {"..", dotdot}};
+  auto dirEntries = lockedDir.getEntries();
+  entries.insert(entries.end(), dirEntries.begin(), dirEntries.end());
+
+#ifdef WASMFS_DEBUG
+  for (auto pair : entries) {
+    emscripten_console_log(pair.name.c_str());
+  }
+#endif
+
+  for (; index < entries.size() && bytesRead + sizeof(dirent) <= count;
+       index++) {
+    auto curr = entries[index];
+    result->d_ino = curr.file->locked().getIno();
+    result->d_off = bytesRead + sizeof(dirent);
+    result->d_reclen = sizeof(dirent);
+    result->d_type =
+      curr.file->is<Directory>() ? DT_DIR : DT_REG; // TODO: add symlinks.
+    strcpy(result->d_name, curr.name.c_str());
```

Comment:
Can we add an assertion that the name fits into the `d_name` field? We also need to enforce that somehow, which I don't think we currently do, although enforcement and testing can be in a separate PR.

Reply:
Done

Reply:
Done